homework/list.js

```diff
@@ -52,28 +45,48 @@
       buff.appendChild(li);
     });
     ul.appendChild(buff);
-    _wrapper.appendChild(ul);
-  }
+    this._wrapper.appendChild(ul);
+  },
 
-  function fetchList(afterFetch) {
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', 'http://127.0.0.1:8000/demo-list-notes.json', true);
-    xhr.responseType = 'json';
-    xhr.onreadystatechange = function(e) {
-      // Watch out: we have a mysterious unknown 'this'.
-      if (this.readyState === 4 && this.status === 200) {
-        var listData = this.response;
-        // The flow ends here.
-        afterFetch(listData);
-      } else if (this.status !== 200 ){
-        // Ignore error in this case.
-      }
-    };
-    xhr.send();
-  }
+  fetchList(url) {
+    return new Promise(function(resolve, reject) {
+      var xhr = new XMLHttpRequest();
+      xhr.open('GET', url, true);
+      xhr.responseType = 'json';
+      xhr.onreadystatechange = (function(e) {
+        // Watch out: we have a mysterious unknown 'this'.
+        if (this.readyState === 4 && this.status === 200) {
+          var listData = this.response;
+          // The flow ends here.
+          resolve(listData);
+        } else if (this.status !== 200 ){
+          reject(new Error('FETCHING FAILED: ' + this.status + ' ' + this.readyState));
+        }
+      }).bind(xhr);
+      xhr.send();
+    });
+  },
 
-  document.addEventListener('DOMContentLoaded', function(event) {
-    start();
-  });
+  handleEvent(event) {
+    switch(event.type) {
+      case 'click':
+        this.onNoteOpen(event);
+        break;
+    }
+  },
+
+  start() {
+    this._wrapper = document.querySelector('#note-list-wrapper');
+    this.fetchList('./demo-list-notes.json').then((function(response) {
+      this.updateList(response);
+      this.drawList();
+      this.preloadFirstNote();
+    }).bind(this)).catch(function(error){
+      console.error(error);
```

Comment:
**Serious issue**: you should **re-throw** the error to make sure the following chains would be interrupted. If you don't do that, the next chain will still perform the actions since the error has already been captured.